File: app/src/main/java/com/tomclaw/appsend_rb/AboutActivity.java
package com.tomclaw.appsend_rb;

import android.content.Intent;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import com.anjlab.android.iab.v3.BillingProcessor;
import com.anjlab.android.iab.v3.Constants;
import com.anjlab.android.iab.v3.TransactionDetails;
import com.google.android.material.snackbar.Snackbar;
import com.tomclaw.appsend_rb.util.ThemeHelper;

import static com.tomclaw.appsend_rb.util.PreferenceHelper.IS_DONATE_ENABLED;

/**
 * Created by Solkin on 17.12.2014.
 */
public class AboutActivity extends AppCompatActivity implements BillingProcessor.IBillingHandler {

    private static final String MARKET_DETAILS_URI = "market://details?id=";
    private static final String MARKET_DEVELOPER_URI = "market://search?q=";
    private static final String GOOGLE_PLAY_DETAILS_URI = "http://play.google.com/store/apps/details?id=";
    private static final String GOOGLE_PLAY_DEVELOPER_URI = "http://play.google.com/store/apps/search?q=";
    public static String DEVELOPER_NAME = "TomClaw Software";

    private BillingProcessor bp;
    private View rootView;
    private View presentButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        ThemeHelper.updateTheme(this);
        super.onCreate(savedInstanceState);

        String licenseKey = getString(R.string.license_key);
        bp = new BillingProcessor(this, licenseKey, this);

        setContentView(R.layout.about_activity);
        ThemeHelper.updateStatusBar(this);

        rootView = findViewById(R.id.root_view);

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        ActionBar actionBar = getSupportActionBar();
        actionBar.setDisplayHomeAsUpEnabled(true);
        actionBar.setDisplayShowHomeEnabled(false);

        TextView appVersionView = findViewById(R.id.app_version);
        PackageManager manager = getPackageManager();
        try {
            PackageInfo info = manager.getPackageInfo(getPackageName(), 0);
            appVersionView.setText(getString(R.string.app_version, info.versionName, info.versionCode));
        } catch (PackageManager.NameNotFoundException ignored) {
        }

        findViewById(R.id.rate_app).setOnClickListener(v -> rateApplication());

        findViewById(R.id.all_apps).setOnClickListener(v -> allProjects());

        presentButton = findViewById(R.id.present_chocolate);
        presentButton.setOnClickListener(view -> onChocolateClicked());

        if ((bp.loadOwnedPurchasesFromGoogle() && bp.isPurchased(getString(R.string.chocolate_id)))
                || !IS_DONATE_ENABLED) {
            presentButton.setVisibility(View.GONE);
        }
    }

    @Override
    public void onDestroy() {
        if (bp != null) {
            bp.release();
        }
        super.onDestroy();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home: {
                finish();
            }
        }
        return true;
    }

    private void onChocolateClicked() {
        String chocolateId = getString(R.string.chocolate_id);
        bp.purchase(this, chocolateId);
    }

    private void rateApplication() {
        final String appPackageName = getPackageName();
        try {
            startActivity(new Intent(Intent.ACTION_VIEW,
                    Uri.parse(MARKET_DETAILS_URI + appPackageName)));
        } catch (android.content.ActivityNotFoundException ignored) {
            startActivity(new Intent(Intent.ACTION_VIEW,
                    Uri.parse(GOOGLE_PLAY_DETAILS_URI + appPackageName)));
        }
    }

    private void allProjects() {
        try {
            startActivity(new Intent(Intent.ACTION_VIEW,
                    Uri.parse(MARKET_DEVELOPER_URI + DEVELOPER_NAME)));
        } catch (android.content.ActivityNotFoundException ignored) {
            startActivity(new Intent(Intent.ACTION_VIEW,
                    Uri.parse(GOOGLE_PLAY_DEVELOPER_URI + DEVELOPER_NAME)));
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (!bp.handleActivityResult(requestCode, resultCode, data)) {
            super.onActivityResult(requestCode, resultCode, data);
        }
    }

    @Override
    public void onProductPurchased(String productId, TransactionDetails details) {
        Toast.makeText(this, R.string.thank_you, Toast.LENGTH_LONG).show();
        presentButton.setVisibility(View.GONE);
    }

    @Override
    public void onPurchaseHistoryRestored() {
    }

    @Override
    public void onBillingError(int errorCode, Throwable error) {
        if (errorCode != Constants.BILLING_RESPONSE_RESULT_USER_CANCELED) {
            Snackbar.make(rootView, R.string.purchase_error, Snackbar.LENGTH_LONG).show();
        }
    }

    @Override
    public void onBillingInitialized() {
    }
}

File: app/src/main/java/com/tomclaw/appsend_rb/screen/about/AboutActivity.kt
package com.tomclaw.appsend_rb.screen.about

import android.content.Context
import android.content.Intent
import android.content.Intent.ACTION_VIEW
import android.net.Uri
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import com.tomclaw.appsend_rb.R
import com.tomclaw.appsend_rb.getComponent
import com.tomclaw.appsend_rb.screen.about.di.AboutModule
import com.tomclaw.appsend_rb.screen.apps.PreferencesProvider
import com.tomclaw.appsend_rb.util.updateStatusBar
import com.tomclaw.appsend_rb.util.updateTheme
import javax.inject.Inject

class AboutActivity : AppCompatActivity(), AboutPresenter.AboutRouter {

    @Inject
    lateinit var presenter: AboutPresenter

    @Inject
    lateinit var preferences: PreferencesProvider

    private var isDarkTheme: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        application.getComponent()
            .aboutComponent(AboutModule(this))
            .inject(activity = this)

        isDarkTheme = updateTheme(preferences)
        updateStatusBar()
        super.onCreate(savedInstanceState)
        setContentView(R.layout.about_activity)

        val view = AboutViewImpl(window.decorView)

        presenter.attachView(view)
    }

    override fun onBackPressed() {
        super.onBackPressed()
        presenter.onBackPressed()
    }

    override fun onStart() {
        super.onStart()
        presenter.attachRouter(this)
    }

    override fun onStop() {
        presenter.detachRouter()
        super.onStop()
    }

    override fun onDestroy() {
        presenter.detachView()
        super.onDestroy()
    }

    override fun openRate() {
        openUriSafe(
            uri = MARKET_URI_RATE + packageName,
            fallback = WEB_URI_RATE + packageName
        )
    }

    override fun openProjects() {
        openUriSafe(
            uri = MARKET_URI_PROJECTS,
            fallback = WEB_URI_PROJECTS
        )
    }

    override fun leaveScreen() {
        finish()
    }

    private fun openUriSafe(uri: String, fallback: String) {
        try {
            startActivity(Intent(ACTION_VIEW, Uri.parse(uri)))
        } catch (ignored: android.content.ActivityNotFoundException) {
            startActivity(Intent(ACTION_VIEW, Uri.parse(fallback)))
        }
    }

}

fun createAboutActivityIntent(
    context: Context
): Intent = Intent(context, AboutActivity::class.java)

private const val MARKET_URI_RATE = "market://details?id="
private const val MARKET_URI_PROJECTS = "market://search?q=pub:TomClaw+Software"
private const val WEB_URI_RATE = "http://play.google.com/store/apps/details?id="
private const val WEB_URI_PROJECTS = "http://play.google.com/store/apps/developer?id=TomClaw+Software"